### PR TITLE
Update snippets so that they can handle custom primary keys.

### PR DESCRIPTION
**Pull request by saevarom**
This is a fix for issue #3104 which describes problems with snippet models having custom primary keys. The wagtailsnippets app mostly assumes an id field everywhere which breaks almost all snippets views and operations.

I've tried to test this as extensively as possible, mostly re-using tests that had been created for the regular snippets, simply replacing the models and fixtures with new ones using custom primary keys. I had to change the URL structure of the snippets app, since the primary keys can include any characters, including `/`. I copied Django's behaviour for quoting and unquoting the primary keys themselves, but found no good way to build the URLs, other than simply allowing all characters (excluding `/`).

Comments and suggestions appreciated :)

### CODE REVIEW

**Comment by saevarom** [comment]
Some tests are still failing. I'll investigate.

File: wagtail/wagtailsnippets/templates/wagtailsnippets/snippets/confirm_delete.html [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {% extends "wagtailadmin/base.html" %}
-{% load wagtailadmin_tags i18n %}
+{% load wagtailadmin_tags i18n admin_urls %}
```
Comment (gasman) [inline wagtail/wagtailsnippets/templates/wagtailsnippets/snippets/confirm_delete.html]: We can't rely on user projects having `django.contrib.admin` in `INSTALLED_APPS`, and if it isn't, the tag library will fail to load. We should probably import or copy the `admin_urlquote` filter into wagtailadmin_tags instead...

**Comment by gasman** [comment]
Rebased (+ tag imports fixed as per https://github.com/wagtail/wagtail/pull/4003#pullrequestreview-86423704) at #4166

**Comment by gasman** [comment]
Merged via #4166 - thanks @saevarom!

**Pull request closed without merging**